src/v/cluster/archival/replica_state_validator.cc

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cluster/archival/replica_state_validator.h"
+
+#include "cloud_storage/types.h"
+#include "cluster/archival/logger.h"
+#include "model/fundamental.h"
+#include "model/offset_interval.h"
+
+namespace archival {
+
+replica_state_validator::replica_state_validator(
+  const storage::log& log, const cloud_storage::partition_manifest& manifest)
+  : _log(&log)
+  , _manifest(&manifest)
+  , _anomalies(validate()) {}
+
+bool replica_state_validator::has_anomalies() const noexcept {
+    return _anomalies.size() > 0;
+}
+
+const std::deque<replica_state_anomaly>
+replica_state_validator::get_anomalies() const noexcept {
+    return _anomalies;
+}
+
+void replica_state_validator::maybe_print_scarry_log_message() const {
+    if (has_anomalies()) {
+        for (const auto& anomalie : _anomalies) {
+            vlog(
+              archival_log.error,
+              "[{}] anomaly detected: {}",
+              _log->config().ntp(),
+              anomalie.message);
+        }
+    }
+}
+
+std::deque<replica_state_anomaly> replica_state_validator::validate() {
+    std::deque<replica_state_anomaly> result;
+    // Before we start uploading we need to make sure that some invariants
+    // are met.
+    // We're validating the replica state over previously uploaded metadata.
+    // The idea is that we can gradually build confidence by making incremental
+    // checks. This code also defines manual intervention point. Once the
+    // problem is detected the NTP archiver shouldn't do anything other than
+    // reporting the problem periodically and asking for intervention. After
+    // examining the anomaly the operator might decide to disable the checks and
+    // allow the uploads to continue. This is an improvement over the situation
+    // when the uploads are getting stuck because we're not doing any extra
+    // work. It may also be easier to diagnose the problem if the archiver is
+    // not running. Potentially, we can do some automatic mitigations
+    // (transferring leadership or blocking writes).
```

Comment:
great comment, but can it go somewhere more accessible to someone looking into what the replica state validator is?